mat64/io.go

```diff
@@ -5,14 +5,180 @@
 package mat64
 
 import (
+	"bytes"
 	"encoding/binary"
+	"errors"
+
+	"github.com/gonum/matrix"
 )
 
-var (
-	littleEndian  = binary.LittleEndian
-	bigEndian     = binary.BigEndian
-	defaultEndian = littleEndian
+// maxLen is the biggest slice/array len one can create on a 32/64b platform.
+const maxLen = int64(int(^uint(0) >> 1))
 
+var (
 	sizeInt64   = binary.Size(int64(0))
 	sizeFloat64 = binary.Size(float64(0))
+
+	errTooBig    = errors.New("mat64: resulting data slice too big")
+	errBadBuffer = errors.New("mat64: data buffer size mismatch")
 )
+
+// MarshalBinary encodes the receiver into a binary form and returns the result.
+//
+// Dense is little-endian encoded as follows:
+//   0 -  7  number of rows    (int64)
+//   8 - 15  number of columns (int64)
+//  16 - ..  matrix data elements (float64)
+//           [0,0] [0,1] ... [0,ncols-1]
+//           [1,0] [1,1] ... [1,ncols-1]
+//           ...
+//           [nrows-1,0] ... [nrows-1,ncols-1]
+func (m Dense) MarshalBinary() ([]byte, error) {
+	bufLen := int64(m.mat.Rows*m.mat.Cols*sizeFloat64 + 2*sizeInt64)
+	if bufLen > maxLen {
```

Comment:
Ok, now I think that I finally (partly) understand. I made a tiny program that tries to allocate a large slice of bytes. When I cross compile with GOARCH=386 and run, the program terminates with unrecoverable panic. When I compile it on amd64 and run, it terminates with a recoverable panic. I don't seem to be able to get the unrecoverable panic on amd64. The length of the slice is 60% of `maxLen`, so I still wonder what is the purpose of this check. On 64bit platforms the condition is never true because we comparse to the _maximum_ integer. On 32bit platforms it can evaluate to true but `make` can obviously crash with slice sizes much lower than maxLen anyway. Please correct me if I am being dense and missing something obvious.